suave/gen/main.go

```diff
@@ -0,0 +1,335 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"go/format"
+	"html/template"
+	"io/ioutil"
+	"sort"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+type fieldInput struct {
+	Name string
+	Typ  abi.Type
+}
+
+type methodElem struct {
+	Name            string
+	Inputs          []fieldInput
+	Outputs         []fieldInput
+	IsComplexOutput bool
+}
+
+type structElem struct {
+	Name   string
+	Fields []fieldInput
+}
+
+type generator struct {
+	Methods []*methodElem
+	Structs []*structElem
+	Hash    string
+}
+
+func (g *generator) encodeMethod(abiMethod abi.Method) {
+	mtd := &methodElem{
+		Name:   abiMethod.Name,
+		Inputs: []fieldInput{},
+	}
+
+	for _, input := range abiMethod.Inputs {
+		g.findStructs(input.Type)
+
+		mtd.Inputs = append(mtd.Inputs, fieldInput{
+			Name: input.Name,
+			Typ:  input.Type,
+		})
+	}
+
+	for _, output := range abiMethod.Outputs {
+		g.findStructs(output.Type)
+
+		mtd.Outputs = append(mtd.Outputs, fieldInput{
+			Name: output.Name,
+			Typ:  output.Type,
+		})
+	}
+
+	g.Methods = append(g.Methods, mtd)
+}
+
+func (g *generator) findStructs(typ abi.Type) {
+	// there can only be structs either in a slice/array or
+	// as sub-elements of another struct
+	if typ.T == abi.TupleTy {
+		name := typ.TupleRawName
+		name = strings.TrimPrefix(name, "Suave")
+
+		// do not encode twice the same struct
+		for _, st := range g.Structs {
+			if st.Name == name {
+				return
+			}
+		}
+
+		elem := &structElem{
+			Name: name,
+		}
+		for indx, field := range typ.TupleElems {
+			elem.Fields = append(elem.Fields, fieldInput{
+				Name: strings.Title(typ.TupleRawNames[indx]),
+				Typ:  *field,
+			})
+			// search for structs inside the struct itself
+			g.findStructs(*field)
+		}
+		g.Structs = append(g.Structs, elem)
+	} else if typ.T == abi.SliceTy || typ.T == abi.ArrayTy {
+		g.findStructs(*typ.Elem)
+	}
+}
+
+func encodeType(typ abi.Type, insideTypes, slicePointers bool) string {
+	// if the type is a tuple, encode as a struct
+	if typ.T == abi.TupleTy {
+		name := typ.TupleRawName
+		name = strings.TrimPrefix(name, "Suave")
+
+		if insideTypes {
+			return name
+		}
+		return "types." + name
+	}
+
+	// if the type is an array, return "[]" + the type of the element
+	if typ.T == abi.SliceTy {
+		if typ.Elem.T == abi.TupleTy && slicePointers {
```

Comment:
This is a hack because the abi package is not able to pack a slice of pointer structs. It will be fixed in a followup PR.